imaging/tools/custom/watershedSegmentationTool.ts

```diff
@@ -0,0 +1,993 @@
+import { each, extend } from "lodash";
+import cv from "@techstark/opencv-js";
+const external = cornerstoneTools.external;
+const BaseBrushTool = cornerstoneTools.importInternal("base/BaseBrushTool");
+const segmentationUtils = cornerstoneTools.importInternal(
+  "util/segmentationUtils"
+);
+const getCircle = segmentationUtils.getCircle;
+const segmentationModule = cornerstoneTools.getModule("segmentation");
+const getToolState = cornerstoneTools.getToolState;
+// internal libraries
+import { DEFAULT_TOOLS } from "../default";
+import store from "../../imageStore";
+const setSegmentationConfig = function (config: Object) {
+  let { configuration } = cornerstoneTools.getModule("segmentation");
+  extend(configuration, config);
+  let enabledElements = cornerstone.getEnabledElements();
+  each(enabledElements, el => {
+    cornerstone.updateImage(el.element);
+  });
+};
+import {
+  mapToRange,
+  getMax,
+  calculateThresholds,
+  shiftAndZeroOut
+} from "./WSUtils";
+import {
+  WSConfig,
+  WSToolConfig,
+  WSMouseEvent,
+  WSEventData,
+  CachedImage,
+  LabelMapType
+} from "../types";
+import { Series } from "../../types";
+
+/**
+ * @public
+ * @class WSTool
+ * @memberof Tools.Brush
+ * @classdesc Tool for drawing segmentations on an image (only pixels inside thresholds)
+ * @extends Tools.Base.BaseBrushTool
+ */
+export default class WSToggleTool extends BaseBrushTool {
+  lowerThreshold: number | null = null;
+  upperThreshold: number | null = null;
+  maskArray: number[][] | null = null;
+  maskArrayCurrentImage: number[] | null = null;
+  src: cv.Mat | null = null;
+  dicomPixelData: number[] | null = null;
+  minThreshold: number | null = null;
+  pixelData: number[][] | null = null;
+  seriesUID: string | null = null;
+  maxThreshold: number | null = null;
+  segmentIndex: number = 1;
+  indexImage: number = 0;
+  imageId: string | null = null;
+  seriesId: string | null = null;
+  labelToErase: number | null = null;
+  click: number = 0;
+  labelToChange: number | null = null;
+  element: HTMLElement | null = null;
+  _lastImageCoords: {
+    x: number;
+    y: number;
+  } | null = null;
+  configuration: WSConfig = {
+    multiImage: false,
+    startIndex: null,
+    endIndex: null,
+    masksNumber: 10
+  };
+  constructor(props = {}) {
+    const defaultProps = {
+      name: "WSToggle",
+      supportedInteractionTypes: ["Mouse", "Touch"],
+      configuration: {
+        multiImage: false,
+        startIndex: null,
+        endIndex: null,
+        masksNumber: 10
+      },
+      mixins: ["renderBrushMixin"]
+    };
+
+    super(props, defaultProps);
+
+    this._handleMouseMove = this._handleMouseMove.bind(this);
+    document.addEventListener("mousemove", this._handleMouseMove);
+    setSegmentationConfig({
+      segmentsPerLabelmap: this.configuration.masksNumber
+    });
+  }
+
+  /**
+   * Allows to get the canvas element when going over it with mouse
+   * TODO check with multiple canvas and layouts
```

Comment:
Todo fatto o ancora da fare ?

Reply:
da testare in dicom vision ma dovrebbe funzionare a regola